### ISSUE TITLE
Feature: control graphql-java errors

### ISSUE THREAD

**Issue by piotrblasiak** [smallrye/smallrye-graphql#776]
Currently (quarkus 1.13.2) graphql-java validation errors are sent without the ability to turn them off:

```
{"errors":[{"message":"The field at path '/model/creationDate' was declared as a non null type, but the code involved in retrieving data has wrongly returned a null value.  The graphql specification requires that the parent field be set to null, or if that is non nullable that it bubble up null to its parent and so on. The non-nullable type is 'DateTime' within parent type 'Model'","path":["model","creationDate"],"extensions":{"classification":"NullValueInNonNullableField"}}],"data":{"model":null}}
```

#772 and #299 are both relevant in this discussion, as it seems to me there should be possible for the developer to exactly specify what messages are sent on errors. I personally don't trust any exception messages not to contain any compromising or useless information to the end-user. In my opinion, only messages that are useful and safe to the end-user should be sent (full developer exception info can be logged instead).

**Comment by t1** [comment]
I can't see why this could be an issue. There's a schema, and either the client or the server don't adhere to it. Validating the schema can't possibly leak anything, can it?

BTW: I understand that some people like to hide the schema, but as long as the field names are not obfuscated as well, I don't see this to be very useful: you can easily guess that e.g. the super hero type will have a name field. Hiding the schema might slow down an attacker by 1%... but it's a big PITA not to have it.

**Comment by piotrblasiak** [comment]
In this case, you are right in that it doesn't leak anything as it simply reacts to fields that are being requested.
But, look at this other message I got:

```
Validation error of type WrongType: argument 'id' with value 'StringValue{value='uZmcZ-obPREWGetAoXxOZ0Q'}' is not a valid 'Unknown Scalar Type [com.myapp.domain.model.GraphId]' @ 'model'
```

I´ve developed the belief that you can't trust exception messages to be sent to the client, as many developers assume their exception messages will stay in the logs. In the message above the class name of the inner workings of my code was leaked, in other cases it might be something more vulnerable.

If we turn it the other way around, besides making it a little bit easier for beginner programmers to find out the problem with their queries, what is the use of sending these exception messages? They don't make any sense to the end-user anyway.

Maybe I am too inexperienced with writing server software, but I am very happy with another API I wrote with graphql-java, where I handle the exceptions and before sending them to the client I make sure only messages I know are safe and understandable to the end-user are sent. The rest are logged and a reference id is sent to the client and shown to the end-user as well.

```
public class ExceptionWhileDataFetching implements GraphQLError {

    private final List<Object> path;
    private final Throwable exception;
    private final List<SourceLocation> locations;
    private final String reference;

    public ExceptionWhileDataFetching(final ResultPath path, final Throwable exception, final SourceLocation sourceLocation, final String reference) {

        this.path = assertNotNull(path).toList();
        this.exception = assertNotNull(exception);
        this.locations = Collections.singletonList(sourceLocation);
        this.reference = reference;
    }

    @Override
    public String getMessage() {
        return exception instanceof UserFriendlyException ? ((UserFriendlyException) exception).getUserFriendlyMessage() : null;
    }
```

Perhaps I am more extreme and paranoid than most, and I would be happy with my way not being standard, but with some way to handle and rethrow any exceptions so that I can satisfy my paranoia and mask any information before it is being sent to the client :)

Once again, great work with the project. Much more enjoyable to work with than straight up graphql-java.

**Comment by t1** [comment]
> besides making it a little bit easier for beginner programmers to find out the problem with their queries

This is not about beginners: the average developer regularly runs into problems with using an API. Directly giving the feedback in the response is extremely helpful. Clients generally don't have access to the logs, so they'll need to contact the developers of the service to check what they where doing wrong. They will instead try to find out what was wrong by reiterating the schema or other docs, searching for their mistake. In a complex schema, this can take literally hours, while a good message would have helped in a second. Having good error messages, i.e. human readable descriptions that pinpoint the problem, is a _huge_ benefit to the clients. There's nothing worse than a REST service just returning a 400 Bad Request without the slightest hint as to what was wrong.

So I think 'leaking' the internal class name `StringValue` is more a question of usability than a security issue.

Only if a service needs an exceptionally high level of security, it may be advisable to limit even this kind of feedback. Such services _have_ to be simple anyways, as complexity is the enemy of security: so it's okay to not have the direct feedback in such cases. But hiding validation error messages is definitively nothing for the average service.

**Comment by piotrblasiak** [comment]
Well, both those errors I posted I would consider Internal server error because there's nothing the client developer can do about them. It's a little besides the point but worth pointing out as I would say that is leaking useless information. 
Anyway, I think there's plenty of arguments for all the different ways to set this up and I buy into your arguments but I don't trust all the libraries involved in a project this big to be written with the awareness that exception messages may leak so in my case since I am full stack anyway I will still go for the more careful option.

Hopefully you agree that one should able to lock it down for extreme security. Being able to register your own exception handler for example would solve this, #299 and #772 .

**Comment by t1** [comment]
I'm perfectly fine with having the option to 'lock it down for extreme security'. I just wouldn't like this to be the default for everyone. Security first, yes, but not Extreme Security first at the cost of usability. And we're only talking about the graphql-java validation errors here: we may come to a different default for the more general case.

**Comment by t1** [comment]
I think that the same reasoning is valid for BeanValidation errors: they are logically a part of the schema, so there's nothing to be leaked.

**Comment by t1** [comment]
I think the leaking aspect has been mitigated by #772. This is still relevant, as we need other things as well. We need a generic mechanism to map exceptions to GraphQL errors.

Then #938 is a duplicate to this.

**Comment by t1** [comment]
I think we can close this. Leaking was fixed with #772 and a generic error mapping mechanism will be implemented in #299
